build-pine64-image.sh

```diff
@@ -23,64 +66,150 @@ if [ "$(id -u)" -ne "0" ]; then
+	fi
 
-# Unpack
-unxz -k --stdout "$SIMPLEIMAGE" > "$TEMP/$IMAGE"
-# Enlarge
-dd if=/dev/zero bs=1M count=$SIZE >> "$TEMP/$IMAGE"
-# Resize
-echo ", +" | sfdisk -N 2 "$TEMP/$IMAGE"
+	echo -n "Checking for presence of device-tree-compiler... "
+
+	if [ -x /usr/bin/dtc ]; then
+		echo "OK"
+	else
+		echo "Not Installed" && DEPENDECIES="1"
+	fi
+	
+	if [ $DEPENDECIES -eq 1 ]; then
+		echo "ERROR: One or more depedecies required to generate the simpleimage are not installed. Please install the missing dependecies and try again or provide a simpleimage for the tool to use"
+		exit 1
+	fi
+
+}
+
+build_simpleimage() {
+	if [ ! -d u-boot-pine64 ]; then
+		git clone https://github.com/longsleep/u-boot-pine64.git
+	fi
+	
+	if [ ! -d arm-trusted-firmware-pine64 ]; then
+		git clone https://github.com/longsleep/arm-trusted-firmware.git arm-trusted-firmware-pine64
+	fi
+
+	if [ ! -d sunxi-pack-tools ]; then
+		git clone https://github.com/longsleep/sunxi-pack-tools.git
+		make -C sunxi-pack-tools
+	fi
+
+	cd u-boot-pine64
+
+	make clean
+	git pull origin
+	make ARCH=arm CROSS_COMPILE=arm-linux-gnueabihf- sun50iw1p1_config
+	make ARCH=arm CROSS_COMPILE=arm-linux-gnueabihf-
+
+	cd ../arm-trusted-firmware-pine64
+
+	make clean
+	git pull origin
+	make ARCH=arm CROSS_COMPILE=aarch64-linux-gnu- PLAT=sun50iw1p1 bl31
+
+	cd ../u-boot-postprocess
+
+	./u-boot-postprocess.sh
+	
+	cd ../simpleimage
+
+	./make_simpleimage.sh $TYPE /tmp/temp-pine64/pine64-image.img $LINUX $SIZE
+	
+}
+
+if [ -z "$SIMPLEIMAGE" ]; then
+	echo "simpleimage is not provided. This tool will now attempt to generate a new simpleimage..."
+	check_dependecies
+	build_simpleimage
+	losetup $LOOP /tmp/temp-pine64/pine64-image.img -o $((143360 * 512))
+else
+	echo "Using the simpleimage from $SIMPLEIMAGE..."
+	# Unpack
+	unxz -k --stdout "$SIMPLEIMAGE" > "/tmp/temp-pine64/pine64-image.img"
+	# Enlarge
+	dd if=/dev/zero bs=1M count=$(($SIZE - 50)) >> "/tmp/temp-pine64/pine64-image.img"
+	# Resize
+	echo ", +" | sfdisk -N 2 "/tmp/temp-pine64/pine64-image.img"
+	losetup $LOOP /tmp/temp-pine64/pine64-image.img -o $((143360 * 512))
+	resize2fs $LOOP
+	cd simpleimage
+fi
 
-# Device
-mkdir "$TEMP/boot"
-mkdir "$TEMP/rootfs"
-DEVICE=$(losetup --show --find "$TEMP/$IMAGE")
-DEVICENAME=$(basename $DEVICE)
-echo "> Device is $DEVICE ..."
-kpartx -avs $DEVICE
+mount $LOOP /tmp/temp-pine64/rootfs
+./make_rootfs.sh /tmp/temp-pine64/rootfs $LINUX $DISTRO
+umount /tmp/temp-pine64/rootfs
+zerofree -v $LOOP
+losetup -d $LOOP
 
-# Resize filesystem
-resize2fs /dev/mapper/${DEVICENAME}p2 || true
+cd ..
 
-# Mount
-mount /dev/mapper/${DEVICENAME}p1 "$TEMP/boot"
-mount /dev/mapper/${DEVICENAME}p2 "$TEMP/rootfs"
+mv /tmp/temp-pine64/pine64-image.img "$PWD/$OUTPUT/$DISTRO2-$TYPE-$(date +%Y%m%d).img"
 
-sleep 2
-(cd simpleimage && ./make_rootfs.sh "$TEMP/rootfs" "$KERNELTAR" "$DISTRO" "$TEMP/boot")
+trap "" EXIT
 
-mv -v "$TEMP/$IMAGE" .
+echo ""
+echo "#"
+echo "# The image has been built successfully"
+echo "#"
+echo ""
```

Comment:
Please remove all these changes. This is not a end user tool and i would very much prefer the script to be as minimal as possible.